### DISCUSSION TITLE
AES Command Line Tools

### DISCUSSION THREAD

**Discussion by wdkaye** [DominoJP/Image-Encryption-AES#7] [General]
I found out that I've already got OpenSSL tools installed on my Mac, from when I was installing GCC using `brew` a few weeks ago for our OpenMP assignment.  After digging around in the manual and googling various options, sure enough, I can encrypt and decrypt files on the command line:

```
$ man openssl

$ openssl aes-128-ecb -in hmmmm.txt -out encrypted

$ openssl aes-128-ecb -d -in encrypted -out unencrypt.txt
```

It prompted me to create a password, and then asked me for it during decryption.

One thing I learned from this was the difference between two versions of AES encryption: ECB (Electronic Cookbook) and CBC (Cipher Blocker Chaining).  In a nutshell, ECB was the first and most basic implementation, while CBC adds complexity by having blocks depend on previous blocks.
https://support.datalocker.com/support/solutions/articles/104691-what-is-the-difference-between-ecb-cbc-mode-and-aes-encryption-

That said, I'm off into the weeds at this point.  I was just looking for info on how to use OpenSSL libraries in my C program, because those are supposedly already installed on my system also:

`#include <openssl/evp.h>`

**Comment by Andrew-Miner** [comment]
Wikipedia has a really good diagram about ECB and CBC: https://upload.wikimedia.org/wikipedia/commons/thumb/e/ef/BlockCipherModesofOperation.svg/2880px-BlockCipherModesofOperation.svg.png

I think the FIPS document you posted only talks about ECB (granted I haven't read through the entire document front to back), which is how I implemented my sequential version.

One thing I'm curious about is how does openssl pad data blocks when they are smaller than 128 bits? The FIPS document does not specify a padding scheme. If we want to use a lib to test with, we will have to use the same padding scheme the lib does. I've read around that PKCS#7 is the most common way to pad, but when I implemented padding in my version I used the ISO/IEC 7816-4 scheme because it was easy and I was tired. It should be simple enough to swap to w/e padding scheme we decide we need.

Edit:
I looked it up and OpenSSL uses PKCS#7 by default. It sounded more complicated when I first read about it but it's actually really simple. Here's a link that describes it: https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS#5_and_PKCS#7

**Reply by wdkaye** [reply]
> One thing I'm curious about is how does openssl pad data blocks when they are smaller than 128 bits? The FIPS document does not specify a padding scheme.

Well I don't have the answer yet, and I was just going to use files that were some multiples of 16 bytes so I didn't have to confront this.  But I was surprised to learn that the encrypted file was larger than my original: my 48-byte text file emerged from the process with an extra 32 bytes added.  I think this has to do with the encryption password that I was prompted to enter...

**Reply by Andrew-Miner** [reply]
Huh that is very strange. I assumed the file size would always be the same if it was divisible by 16.

Edit: If you haven't already you could try adding the -nopad option to stop openssl from padding and see if it changes anything

**Reply by DominoJP** [reply]
That extra 32 bytes sounds like it might be due to padding and possibly some metadata being added during encryption. As Andrew mentioned, OpenSSL uses PKCS#7 padding by default, which seems to ensure that the block size aligns with the encryption algorithm. If the data is a perfect multiple of the block size), then PKCS#7 still adds an entire block of padding to distinguish cases where no padding would be needed? I think I am understanding this correctly.

**Reply by Andrew-Miner** [reply]
Yea thats probably it, the wiki page is super confusing in the wording.

> Consider a plaintext message that is an integer multiple of B bytes with the last byte of plaintext being 01. With no additional information, the deciphering algorithm will not be able to determine whether the last byte is a plaintext byte or a pad byte. However, by adding B bytes each of value B after the 01 plaintext byte, the deciphering algorithm can always treat the last byte as a pad byte and strip the appropriate number of pad bytes off the end of the ciphertext; said number of bytes to be stripped based on the value of the last byte. 

Sounds like no matter the file size there is going to be some padding. Weird thats its 32 bytes extra though and not 16, since AES uses 16 byte blocks.

Edit: 

OK I think I figured out why warren's file was 32 bytes longer. 16 bytes were from the padding. When the file is evenly divisible by 16, 16 bytes of padding get added. I think the second 16 bytes comes from the salt feature. I don't fully understand what it does yet but when you do -nosalt with openssl that last extra 16 bytes disappears.

Still I havent managed to get my implementation to match an OpenSSL encrypted file yet. It matches some online web tools though
